### ISSUE TITLE
Proposal: unify rust bindings for mesos

### ISSUE THREAD

**Issue by ConnorDoyle** [bonifaido/resos#12]
Hi there!

I just came across this fork and it looks like some of the history is based on a first attempt I made some time ago.  Nice work so far with the HTTP support.  I ended up trashing the `resos` library and started over here, exclusively working on the "native" (linking to libmesos) bindings: https://github.com/ConnorDoyle/mesos-rust.  The scheduler can launch, kill, and get status of tasks.  The native executor driver is a work in progress but not far from done.  I tried to keep the driver code that interacts with the native library in a separate package (`native`) so that the same scheduler and executor interfaces could be used with the HTTP or libmesos backed driver.  The interfaces in mesos-rust and the C shims need to be updated to 0.27 (I'm happy to do that).  I'm curious whether you think it's worthwhile to join forces and combine these libraries into one project.

Thanks for reading this far.

**Comment by ConnorDoyle** [comment]
PS, if you are open to this we could create a shared org on GitHub and both have write access.

**Comment by bonifaido** [comment]
Hi!

Absolutely, I mean great idea, let's do that. I think creating a GitHub org for this, is the way to go. I just started to lay down the interface for the Executors as well, but ran out of time, so this will be a nice place to continue this effort! Do you mind creating this organization? Than we should discuss how to merge the two projects into one.
I'm working on a rusty zookeeper library as well, which I planed to use for the leader election here.

**Comment by ConnorDoyle** [comment]
Done: https://github.com/iron-oxide
I added you to the org, looks like as soon as you accept the invite you will become an org owner and a member of the @iron-oxide/maintainers team.

Regarding how to merge, here's one idea (open to others): _mesos-rust_ has a `native` package where the _libmesos_ backed drivers live.  Maybe we could add an `http` package for the HTTP-API backed drivers.  We'd have to update the scheduler and executor traits a bit but that should be done anyway.

**Comment by bonifaido** [comment]
I have joined the organization, and I started to check out the initiating project (yes, my version was originally a fork of yours :-) ). I ran into some issues when compiling it, namely with protobuf versions, which version do you use? This is how I wanted to build, I guess there is something with my protobuf version:

``` shell
brew install mesos protobuf # mesos 0.27.2, protobuf 2.6.1
cd mesos-rust
make
```

And the errors (many of this):

```
/usr/local/include/mesos/mesos.pb.h:10994:48: error: no member named 'kEmptyString' in namespace 'google::protobuf::internal'
  if (value_ != &::google::protobuf::internal::kEmptyString) {
```

I think I need protobuf 2.5 (I see native mesos protobuf classes are generated with this version).

**Comment by ConnorDoyle** [comment]
Yep, Mesos is still on 2.5, guess I should call that out in the build instructions.

**Comment by ConnorDoyle** [comment]
Also, you'll need libprotobuf-lite on the linker path (not sure if homebrew installs that). I build the protobuf libs from source on my machine.

**Comment by bonifaido** [comment]
Okay, I have started to integrate the HTTP implementation into the project, but ran into some issues with the protobuf messages, I think they are older (you have mentioned this already if I'm right). I have a script in the proto directory which did the job for me!

**Comment by ConnorDoyle** [comment]
I will update the protobufs, generated code, and C bindings to `0.28.0` (released last week) over the weekend.  Thanks for the pointer to your script.

**Comment by ConnorDoyle** [comment]
@bonifaido, finally got to this (I have a good excuse -- new baby! 👶 ).
The PR is here: https://github.com/iron-oxide/mesos-rust/pull/19

**Comment by ConnorDoyle** [comment]
That PR only updates the generated code and cleans up the build a little.  I'll continue to work on updating the C shim and scheduler / executor traits to be consistent with the latest API.

**Comment by bonifaido** [comment]
Congratulations! 👍  What a coincidence, we got our new baby at the end of January. :)

What I have discovered so far is that the native API and the HTTP API are divergent a bit, at least in lingo, because I couldn't compile the http code with the protos inside mesos-rust, for example SlaveID is AgentID in http. I wonder what is the reason behind this.

Until that I develop the driver here, to keep it up to update with 0.28.1, so we can merge easily later.
